Objects/unionobject.c

```diff
@@ -108,25 +108,282 @@ static PyMethodDef union_methods[] = {
-        return PyObject_RichCompare(aa->args, bb->args, Py_EQ);
+    if (type == &Py_UnionType) {
+        unionobject *bb = (unionobject *)b;
+        b_set = PySet_New(bb->args);
+    }
+
+    if (b_set == NULL) {
+        Py_RETURN_FALSE;
     }
-    Py_RETURN_FALSE;
+
+    return PyObject_RichCompare(a_set, b_set, Py_EQ);
 }
 
+static PyObject*
+flatten_args(PyObject* args) {
+    int arg_length = PyTuple_GET_SIZE(args);
+    int total_args = 0;
+    // Get number of args once it's flattened.
+    for (int i = 0; i < arg_length; i++) {
+        PyObject *arg = PyTuple_GET_ITEM(args, i);
+        PyTypeObject* arg_type = Py_TYPE(arg);
+        if (arg_type == &Py_UnionType) {
+            unionobject *alias = (unionobject *)arg;
+            total_args = total_args + PyTuple_GET_SIZE(alias->args);
+        } else {
+            total_args++;
+        }
+    }
+    // Create new tuple of flattened args.
+    PyObject *flattened_args = PyTuple_New(total_args);
+    int pos = 0;
+    for (int i = 0; i < arg_length; i++) {
+        PyObject *arg = PyTuple_GET_ITEM(args, i);
+        PyTypeObject* arg_type = Py_TYPE(arg);
+        if (arg_type == &Py_UnionType) {
+            unionobject *alias = (unionobject *)arg;
+            PyObject* nested_args = alias->args;
+            int nested_arg_length = PyTuple_GET_SIZE(nested_args);
+            for (int j = 0; j < nested_arg_length; j++) {
+                PyObject* nested_arg = PyTuple_GET_ITEM(nested_args, j);
+                PyTuple_SET_ITEM(flattened_args, pos, nested_arg);
+                pos++;
+            }
+        } else {
+            PyTuple_SET_ITEM(flattened_args, pos, arg);
+            pos++;
+        }
+    }
 
+    return flattened_args;
+}
+
+static PyObject*
+dedup_and_flatten_args(PyObject* args) {
+    args = flatten_args(args);
+    int arg_length = PyTuple_GET_SIZE(args);
+    PyObject* temp[arg_length];
+    for (int i = 0; i < arg_length ; i++) { temp[i] = NULL; }
+
+    // Add unique elements to an array.
+    int temp_count = 0;
+    for(int i = 0; i < arg_length; i++) {
+        int is_duplicate = 0;
+        PyObject* i_tuple = PyTuple_GET_ITEM(args, i);
+        for(int j = i + 1; j < arg_length; j++) {
+            PyObject* j_tuple = PyTuple_GET_ITEM(args, j);
+            if (i_tuple == j_tuple) is_duplicate = 1;
+        }
+        if (!is_duplicate) {
+            temp[temp_count] = i_tuple;
+            temp_count++;
+        }
+    }
+
+    // Create new tuple from deduplicated array.
+    PyObject* temp_arg = NULL;
+    PyObject* new_args = PyTuple_New(temp_count);
+    for(int k = 0; k < arg_length; k++) {
+        temp_arg = temp[k];
+
+        if (temp_arg != NULL)
+            PyTuple_SET_ITEM(new_args, k, temp_arg);
+    }
+
+    return new_args;
+}
+
+
+static int
+is_typevar(PyObject *obj)
+{
+    return is_typing_name(obj, "TypeVar");
+}
+
+static int
+is_genericalias(PyObject *obj)
+{
+    return is_typing_name(obj, "GenericAlias");
```

Comment:
This probably doesn't work, because `typing.GenericAlias` is just a coincidental alias for `types.GenericAlias`. (Check it in the REPL.)